src/mono/mono/mini/interp/interp.c

```diff
@@ -1442,115 +1464,136 @@ build_args_from_sig (MonoMethodSignature *sig, InterpFrame *frame)
 		case MONO_TYPE_CLASS:
 		case MONO_TYPE_OBJECT:
 		case MONO_TYPE_STRING:
-#if SIZEOF_VOID_P == 8
 		case MONO_TYPE_I8:
 		case MONO_TYPE_U8:
+		case MONO_TYPE_VALUETYPE:
+		case MONO_TYPE_GENERICINST:
+			info->ret_type = PINVOKE_ARG_INT;
+			break;
+		case MONO_TYPE_R4:
+		case MONO_TYPE_R8:
+			info->ret_type = PINVOKE_ARG_R8;
+			break;
+		case MONO_TYPE_VOID:
+			info->ret_type = PINVOKE_ARG_NONE;
+			break;
+		default:
+			g_error ("build_args_from_sig: ret type not implemented yet: 0x%x\n", sig->ret->type);
+	}
+
+	return info;
+}
+
+static void
+build_args_from_sig (InterpMethodArguments *margs, MonoMethodSignature *sig, BuildArgsFromSigInfo *info, InterpFrame *frame)
+{
+#ifdef TARGET_ARM
```

Comment:
Not sure how this is the first time I'm noticing arm handling here. The `build_args_from_sig` path is wasm only. All other platforms have pinvoke tramp. Some of the code from this PR could be removed/simplified.